core/src/main/java/com/rudderstack/android/sdk/core/RudderContext.java

```diff
@@ -65,14 +66,19 @@ public class RudderContext {
         // cachedContext is used every time, once initialized
     }
 
-    RudderContext(Application application, String anonymousId, String advertisingId, String deviceToken) {
+    RudderContext(Application application, String anonymousId, String advertisingId, String deviceToken, boolean collectDeviceId) {
         RudderPreferenceManager preferenceManger = RudderPreferenceManager.getInstance(application);
 
         if (TextUtils.isEmpty(anonymousId)) {
-            anonymousId = preferenceManger.getAnonymousId() != null ? preferenceManger.getAnonymousId() : Utils.getDeviceId(application);
-        } else {
-            preferenceManger.saveAnonymousId(anonymousId);
+            anonymousId = preferenceManger.getAnonymousId();
+            if (anonymousId == null && collectDeviceId) {
+                anonymousId = Utils.getDeviceId(application);
+            }
+            if (anonymousId == null) {
+                anonymousId = UUID.randomUUID().toString();
+            }
         }
+        preferenceManger.saveAnonymousId(anonymousId);
```

Comment:
Please note this change. Moving this out of the "else" block is a suggestion to prevent anonymousIds from being unique on every app launch.
This makes more sense to me, but maybe you put it in the else for a good reason I haven't thought of.

Reply:
it should only affect the cases where the deviceId was unretrievable or collectDeviceId is `false`.